### PR TITLE
Set up a Continuous Integration system for all pull-requests #3295

### LINKED ISSUE #3295: Set up a Continuous Integration system for all pull-requests
1. It's time-consuming (~30min) for reviewers to run all test cases to ensure dev green before merge a pull-request.
2. Instead, it's professional to use Continuous Integration (CI, eg Jenkins) to monitor our PR list. If there's new PR, it will pull the PR, build, run all tests and run failed test against the master branch for some times.
3. It can even do more, eg to build using Maven with Cobertura plugin to ensure that test coverage % is always above certain stat, with PMD to find common programming flaws, verify coding & naming styles etc.
4. When the test is dev green, CI can post a status (green or not) and result link back to the PR. Thus, the reviewers only need to focus on those green ones and if codes are fine, we can directly merge it.

It's feasible, super productive, and I think @thyageshm has done half of the work already.

### PR DESCRIPTION
**Pull request by ashrayjain**
Fixes #3295 

It's finally here!! :)

I have updated the docs but some things may have slipped out. Please let me know if so.
Also I am wondering if we need a separate "How to use gradle" doc to document all the functions in the script?

### CODE REVIEW

**Comment by damithc** [comment]
Nice. I'm going to give it a try soon. Yes, I think we need some documentation on using gradle tasks.

**Comment by damithc** [comment]
This may be incorrect.
image

**Comment by ashrayjain** [comment]
whoops. will fix that.

**Comment by damithc** [comment]
I'm trying it on Eclipse Mars. If it works out, we can update the setup guide to `Luna or Mars`

**Comment by damithc** [comment]
Gradle should be included in the tool stack right?

**Comment by damithc** [comment]
Need to explain how to deal with this warning:
image

**Comment by ashrayjain** [comment]
Will update that also. This warning will only appear for existing devs. New environments won't have that.
You can click on "Overwrite". This basically removes references to the old project.

File: devdocs/process.md [hunk 1]
```diff
@@ -146,7 +155,7 @@ This workflow is an adaptation of the [GitHub flow](https://guides.github.com/in
 ###Reviewing a fix
 Role: reviewer
 
-  * This is a code quality review. No need to run tests.
+  * Ensure that the Travis CI build is successful and the developer has local dev green.
```
Comment (wkurniawan07) [inline devdocs/process.md]: Does this mean that we now have to ensure _two_ kinds of green instead of one? It seems like Travis is adding an overhead here.

Comment (damithc) [inline devdocs/process.md]: I think local dev green means those few tests that cannot be run on Travis.

Comment (ashrayjain) [inline devdocs/process.md]: Yes, thats right.

**Comment by damithc** [comment]
Regarding the warning, I was using a fresh eclipse and a fresh clone of repo, so I was like a new developer, and yet I got that warning.

**Comment by belyabl9** [comment]
Hi,

This is really great enhancement. It will make developing for TEAMMATES much more convenient.
Could you clarify about "localTests", what are they ?
What tests are going to be run locally ? I am mostly interested in ui-browser-tests, as they are the most expensive for time.

**Comment by ashrayjain** [comment]
@belyabl9 "localTests" will be a small subset of all the tests. They will include some chosen tests (which may include a few ui browser tests) that will function as smoke tests and some tests that are unstable on Travis. This PR does not really choose the smoke tests. I think that will be done as a different PR.

Ideally, there would be no unstable tests, but we still have that category to account for tests that may be added and may not be stable on Travis right away (all unstable tests should eventually move to stable).

**Comment by wkurniawan07** [comment]
@belyabl9 all of us in the core team are as interested as you are with this enhancement. Currently we still have some loose ends to tie, we will roll out the details when we're ready to so please be patient :+1:

**Comment by belyabl9** [comment]
Thanks for the information. I will be patient =) I did not want to rush somebody with answers, these are just questions which I have not got from commit files)

**Comment by ashrayjain** [comment]
@damithc @wkurniawan07 Builds are failing after #5058 
Debugging shows text like ```Alice Betsy</option>></td></div>``` etc being displayed to the user. I have attached a screenshot from travis too.
screen shot 2016-03-26 at 8 15 10 pm

From Travis, there are 4 tests that are failing:
1. FeedbackRubricQuestionUiTest
2. InstructorFeedbackResultsPageUiTest
3. FeedbackRankQuestionUiTest
4. StudentFeedbackSubmitPageUiTest

Not sure whats going on. Interestingly, it seems to pass on local testing.

**Comment by wkurniawan07** [comment]
Hmmm the Travis report only shows `FeedbackRubricQuestionUiTest` failing though?

**Comment by ashrayjain** [comment]
Oh! yeah, i ran a single test on travis for debugging. The last few builds (the ones that completed in ~5 mins) have just one test. Look at the older ones (~19 min runtime).

**Comment by damithc** [comment]
I noticed it too in the mashup test. If these extra html text were inserted to test sanitization, it should be made clear in the UI itself. Otherwise the UI looks like broken. @bamboo3250

**Comment by ashrayjain** [comment]
@wkurniawan07 @damithc are you able to get these tests passing locally on master?

**Comment by damithc** [comment]
@bamboo3250 merged a PR a few hours ago. May be he can confirm if master is dev green.

**Comment by bamboo3250** [comment]
Yes, I merged 2 PRs in this afternoon and both are dev-green. Maybe we can try merge master into this branch?

**Comment by ashrayjain** [comment]
i will rebase master and try again.

**Comment by ashrayjain** [comment]
@bamboo3250 is having those strings like ```Alice Betsy</option></td></div>``` in the UI expected?

**Comment by bamboo3250** [comment]
Yes, I purposely made them like that so when those strings are not sanitized properly, it will break the UI by matching with one of the opening tags and we can see that easily. 
And here, by looking at the indentation, the UI is not broken so it should be fine. But unfortunately, the browser automatically renders all sanitized strings like `&lt;` and `&gt;` into `<` and `>` so we cannot see it clearly.

**Comment by damithc** [comment]
@bamboo3250 If this is expected behavior, tweak the data to make it obvious so that the person eyeballing the UI tests don't mistake this as a UI problem. e.g. add a test `some html tags should appear here`. By the I prefer if the testing of sanitizing is done separately rather than affecting other unrelated tests. At the moment they even appear in the mashup pages test which is used to eyeball the UI for problems.

**Comment by ashrayjain** [comment]
@damithc @bamboo3250 After rebasing on master, I can't get the test to pass on Travis or locally. Could you confirm that this test passes for you on master locally?

**Comment by bamboo3250** [comment]
I tried `FeedbackRubricQuestionUiTest`. It passes in my local server (on master).

image

**Comment by ashrayjain** [comment]
@bamboo3250 and just to confirm, you are using Firefox 38.0.5 for testing?

**Comment by bamboo3250** [comment]
@ashrayjain No, I am not. That could be the reason. Let me try again.

**Comment by ashrayjain** [comment]
@bamboo3250  yup, let me know if it passes on that as well. thanks! :)

**Comment by bamboo3250** [comment]
@ashrayjain sure. I will ping you soon.

**Comment by bamboo3250** [comment]
I installed FireFox 38.0.5 and ran all tests again and it was still dev-green. I looked at the diff in Travis and realized there is something wrong in Godmode. I will fix it and try generating HTML output again.

**Comment by ashrayjain** [comment]
@bamboo3250 yup, somethings wrong there. I tried printing out expected and actual strings before the comparison is done. The actual strings have the escaped characters but expected doesn't. For example, the actual string contains stuff like ```Alice Betsy&lt;/option&gt;&lt;/td&gt;&lt;/div&gt;'"``` whereas the expected is ```Alice Betsy</option></td></div>'"```

**Comment by damithc** [comment]
It's still interesting why it passes on local but fails on travis.

**Comment by damithc** [comment]
@ashrayjain this is a separate issue: on my Win8 machine, the vm started by gradle (2 java processes) lives on even after closing the cmd window. This is a problem because the task `clean` fails probably because the previous gradle process is not allowing to delete some of the files.

**Comment by bamboo3250** [comment]
@damithc @ashrayjain hopefully we will figure out why it fails on Travis after I merge #5143

**Comment by ashrayjain** [comment]
@damithc actually its failing locally also on Windows for me. @bamboo3250 what dev env are you working with? Mac or Windows?

**Comment by ashrayjain** [comment]
@damithc That is expected. Gradle runs in daemon mode, which mean it lives on even after the task is done. This speeds up subsequent task runs. We need this to allow appengine server to run in background

**Comment by bamboo3250** [comment]
@ashrayjain I'm running on Mac :worried:

**Comment by ashrayjain** [comment]
I see. Yup, its passing for me on Mac but not on Windows. Does anyone have a Windows environment setup using Eclipse to verify this?

**Comment by bamboo3250** [comment]
@kanghj Is the current master passing for you?

**Comment by damithc** [comment]
I tried these two tests on Windows and they are passing on latest master:
FeedbackRubricQuestionUiTest
StudentFeedbackSubmitPageUiTest

**Comment by damithc** [comment]
In #5138 a dev is reporting a test failure related to this.

**Comment by bamboo3250** [comment]
@damithc I suspect it is due to the usage of HashMap. When we get the list of entries in HashMap, they might not be sorted and sometimes there are differences between Mac and Windows.

**Comment by bamboo3250** [comment]
Or that could be just a fault in sorting, where we didn't define the comparing method clearly.

**Comment by ashrayjain** [comment]
@bamboo3250 Good new! I manually added the test html files from #5143 and that seems to fix the problem. Waiting eagerly for it to be merged :)

**Comment by ashrayjain** [comment]
@damithc Also, I figured out the reason it passed locally but not on Travis. Our comparison function seems to have multiple steps. The problem being fixed by #5143 affects one of the latter steps. On local env, the comparison passes on the first try itself, but on Travis it takes a couple of tries and as soon as it fails once, the buggy transformation being fixed gets applied (breaking the expected html).

**Comment by damithc** [comment]
On my Win8, if I start app engine, then stop it, and then clean, it fails with this message.
> Unable to delete file: C:\teammates-2016Mar\build\exploded-app\WEB-INF\lib\appengine-api-1.0-sdk-1.9.34.jar

I have to go to task manager and kill all the java background processes to get the clean to work. Not sure it's specific to my comp or Win8.

File: src/test/java/teammates/test/cases/ui/browsertests/StudentProfilePageUiTest.java [hunk 2]
```diff
@@ -284,7 +284,7 @@ private StudentProfilePage getProfilePageForStudent(String studentId) {
     }
 
     @AfterClass
-    public void testAfter() {
+    public static void classTearDown() throws Exception {
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/browsertests/StudentProfilePageUiTest.java]: Out of curiosity, is this method static as required by Travis, or just to make it similar to other UI tests?

Comment (ashrayjain) [inline src/test/java/teammates/test/cases/ui/browsertests/StudentProfilePageUiTest.java]: Yea, this was to make it uniform.

**Comment by wkurniawan07** [comment]
@ashrayjain you can include `InstructorFeedbackEditPageUiTest` now, it should be working. At least if it still fails it's no longer because of the datepicker focus thingy.

**Comment by ashrayjain** [comment]
@wkurniawan07 :+1: Works great now!

**Comment by wkurniawan07** [comment]
I've been following this for quite some time -- seems like there's one naughty test that keeps failing now?

**Comment by ashrayjain** [comment]
@wkurniawan07 yup, testViewPhotoAndAjaxForLargeScaledSession in InstructorFeedbackResultsPageUiTest is failing pretty consistently now...

**Comment by wkurniawan07** [comment]
Errr... is it really a good idea to merge this now? There are some parts that I haven't reviewed yet...

**Comment by damithc** [comment]
Yes, the PR seemed to have skipped the usual review process. The merged version is not as disruptive as we planned at first. So this early merge will give us more time to dogfood it before exam starts. Meanwhile, you can still discuss the PR here and ask for changes, which can do done as new PRs.

**Comment by wkurniawan07** [comment]
@ashrayjain question: how do you test multiple merges in Travis without merging to `master` first? It's not documented in `process.md`.
What I did previously was merging to a random branch, adding that branch to `.travis.yml`, and push to trigger build. If it passes, then `master` will just fast-forward. Is there another way that's more commonly used?

**Comment by ashrayjain** [comment]
@wkurniawan07 
good question! unfortunately, doing this is not as straightforward with CI. So, one option is doing merges one at a time only and merging master into the feature branch before running on Travis. The Travis build will then have code from both master and the feature branch. This, however, is not very efficient.

In the future, we could explore moving to some variant of another popular git flow (https://gist.github.com/djspiewak/9f2f91085607a4859a66) where there is another branch (called DMZ in the given link) where PRs are merged and tested, before merging to master.

Another alternative to the first option could be automating the merging of master to the feature branch as part of the build itself. This would be a process local to Travis (the merge happens on Travis only) and would not affect the Github's versions. So, basically, the Travis build for every PR would run the build on not only the PR's version of the code but also a version where the PR's code is merged with master. But again, all of this should probably be done at a later time.

**Comment by damithc** [comment]
Not that it matters,  but we can increase the effort label to reflect research effort.

**Comment by damithc** [comment]
Sorry, I mixed up the effort for issue vs review.

**Comment by ashrayjain** [comment]
@wkurniawan07 Upon further investigation, it looks like Travis already builds a version that is basically the PR merged into master!! :) This is super cool. A green Travis build for a PR signifies a green travis build after merging the PR into master.

**Comment by damithc** [comment]
Cool. Does that mean when master is updated, all open PRs are rebuilt?

**Comment by ashrayjain** [comment]
@damithc Nope, that doesn't seem to be supported for efficiency reasons (kinda makes sense when you consider the overhead for highly active projects).
More discussion on this @ https://github.com/travis-ci/travis-ci/issues/1620

A simple workaround could be to re-run the build for a PR once before merging. (especially when the last build for the PR is quite old)

More complicated workarounds would involve the Travis API and us building this feature ourselves! But we would probably run into concurrent build limits leading to increase wait (and build) times.
